File: cliente/AutosCliente/Controllers/BateriaController.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using AutosCliente.Models;
using AutosCliente.Services;

namespace AutosCliente.Controllers
{
    // Controlador para Bateria (clase D)
    public class BateriaController
    {
        private readonly AutoService _service;

        public BateriaController(AutoService service)
        {
            _service = service;
        }

        public async Task<(bool ok, string mensaje, Bateria? resultado)> Agregar(
            string tipo, string capacidad, string ciclos)
        {
            if (!double.TryParse(capacidad, out double capD)) return (false, "Capacidad inválida.", null);
            if (!int.TryParse(ciclos, out int ciclosI)) return (false, "Ciclos inválidos.", null);

            var b = new Bateria { Tipo = tipo.Trim(), CapacidadKwh = capD, CiclosVida = ciclosI };
            var r = await _service.AgregarBateria(b);
            return r != null ? (true, $"Batería agregada. ID: {r.Id}", r) : (false, "Error al agregar.", null);
        }

        public async Task<(bool ok, string mensaje, Bateria? resultado)> Buscar(string idStr)
        {
            if (!int.TryParse(idStr, out int id)) return (false, "ID inválido.", null);
            var b = await _service.BuscarBateriaPorId(id);
            return b != null ? (true, "", b) : (false, "Batería no encontrada.", null);
        }

        public async Task<(bool ok, string mensaje, Bateria? resultado)> Actualizar(
            int id, string tipo, string capacidad, string ciclos)
        {
            if (!double.TryParse(capacidad, out double capD)) return (false, "Capacidad inválida.", null);
            if (!int.TryParse(ciclos, out int ciclosI)) return (false, "Ciclos inválidos.", null);

            var b = new Bateria { Tipo = tipo.Trim(), CapacidadKwh = capD, CiclosVida = ciclosI };
            var r = await _service.ActualizarBateria(id, b);
            return r != null ? (true, "Batería actualizada.", r) : (false, "Error al actualizar.", null);
        }

        public async Task<(bool ok, string mensaje)> Eliminar(int id)
        {
            bool ok = await _service.EliminarBateria(id);
            return ok ? (true, "Batería eliminada.") : (false, "Error al eliminar.");
        }

        public Task<List<Bateria>?> ListarTodas() => _service.ListarBaterias();
    }
}

File: cliente-js/controllers/BateriaController.js
import { autoService } from '../services/AutoService.js';
import { BateriaBuilder } from '../models/Bateria.js';

export class BateriaController {

  async agregarBateria(tipo, capacidad, ciclos, fecha, estado) {
    const capD = parseFloat(capacidad), ciclosI = parseInt(ciclos);
    if (!tipo.trim() || isNaN(capD) || isNaN(ciclosI) || !fecha || !estado.trim())
      return { ok: false, mensaje: 'Todos los campos son obligatorios.' };

    const b = new BateriaBuilder()
      .setTipo(tipo.trim()).setCapacidadKwh(capD).setCiclosVida(ciclosI)
      .setFechaInstalacion(fecha).setEstado(estado.trim()).build();

    const r = await autoService.agregarBateria(b);
    return r ? { ok: true, mensaje: `Batería agregada. ID: ${r.id}` } : { ok: false, mensaje: 'Error al agregar.' };
  }

  async buscarBateria(id) {
    const idI = parseInt(id);
    if (isNaN(idI)) return { ok: false, mensaje: 'ID inválido.' };
    const b = await autoService.buscarBateriaPorId(idI);
    return b ? { ok: true, data: b } : { ok: false, mensaje: 'Batería no encontrada.' };
  }

  async actualizarBateria(id, tipo, capacidad, ciclos, fecha, estado) {
    const b = new BateriaBuilder();
    if (tipo) b.setTipo(tipo.trim());
    if (capacidad) b.setCapacidadKwh(parseFloat(capacidad));
    if (ciclos) b.setCiclosVida(parseInt(ciclos));
    if (fecha) b.setFechaInstalacion(fecha);
    if (estado) b.setEstado(estado.trim());
    const built = b.build();

    const r = await autoService.actualizarBateria(id, built);
    return r ? { ok: true, mensaje: 'Batería actualizada.' } : { ok: false, mensaje: 'Error al actualizar.' };
  }

  async eliminarBateria(id) {
    const ok = await autoService.eliminarBateria(id);
    return ok ? { ok: true, mensaje: 'Batería eliminada.' } : { ok: false, mensaje: 'Error al eliminar.' };
  }

  async listarBaterias() { return await autoService.listarBaterias(); }
}
